### PR TITLE
refactor: enable -Wconversion warning for spell.c

### PR DESCRIPTION
**Pull request by dundargoc**
Work on https://github.com/neovim/neovim/issues/567

### CODE REVIEW

File: src/nvim/spell.c [hunk 1]
```diff
@@ -137,7 +137,7 @@ slang_T *first_lang = NULL;
 char_u *int_wordlist = NULL;
 
 typedef struct wordcount_S {
-  uint16_t wc_count;                // nr of times word was seen
+  int wc_count;                     // nr of times word was seen
```
Comment (zeertzjq) [inline src/nvim/spell.c:140]: ?

Comment (dundargoc) [reply src/nvim/spell.c:140]: It's the simplest way to take care of the conversion warning in line 1899:

```c
if ((wc->wc_count += count) < (unsigned)count) {    // check for overflow
```

Comment (bfredl) [reply src/nvim/spell.c:140]: maybe just `(wc->wc_count += (uint16_t)count)` ? No problem if that truncates the value, the if condition is checking for that as well.

Comment (dundargoc) [reply src/nvim/spell.c:140]: That will still give a conversion warning. Even if both variables are `uint16_t` the sum will still be promoted to `int`. I'll either need to change the type like I did or rewrite the if statement.

File: src/nvim/spell.c [hunk 2]
```diff
@@ -1896,11 +1892,12 @@ void count_common_word(slang_T *lp, char_u *word, int len, int count)
...
     wc = xmalloc(sizeof(wordcount_T) + p_len);
     memcpy(wc->wc_word, p, p_len + 1);
-    wc->wc_count = count;
+    wc->wc_count = (uint16_t)count;
     hash_add_item(&lp->sl_wordcount, hi, wc->wc_word, hash);
   } else {
     wc = HI2WC(hi);
-    if ((wc->wc_count += count) < (unsigned)count) {    // check for overflow
+    wc->wc_count = (uint16_t)(wc->wc_count + count);
+    if (wc->wc_count < count) {    // check for overflow
```
Comment (zeertzjq) [inline src/nvim/spell.c:1900]: It may be better to change the type of `count` to `uint16_t`, and keep this `unsigned` cast.

Comment (dundargoc) [reply src/nvim/spell.c:1900]: Wait what?

So the line should be

```c
    wc->wc_count = (uint16_t)(wc->wc_count + (unsigned)count);
```
?

Comment (zeertzjq) [reply src/nvim/spell.c:1900]: I mean the argument `count`

File: src/nvim/spell.c [hunk 3]
```diff
@@ -1900,7 +1896,8 @@ void count_common_word(slang_T *lp, char_u *word, int len, int count)
...
```
Comment (zeertzjq) [inline src/nvim/spell.c:1900]: I think this can now be changed back.

Comment (dundargoc) [reply src/nvim/spell.c:1900]: That will give a conversion error, so I think not.

File: src/nvim/spell.c [hunk 4]
```diff
@@ -4266,14 +4259,13 @@ static void suggest_trie_walk(suginfo_T *su, langp_T *lp, char_u *fword, bool so
...
-                        fword + sp->ts_fidx, l);
-                sp->ts_prewordlen += l;
+                memmove(preword + sp->ts_prewordlen, fword + sp->ts_fidx, (size_t)l);
+                sp->ts_prewordlen = (char_u)(sp->ts_prewordlen + l);
                 preword[sp->ts_prewordlen] = NUL;
               } else {
                 sp->ts_score -= SCORE_SPLIT - SCORE_SUBST;
               }
-              sp->ts_fidx += l;
+              sp->ts_fidx = (char_u)(sp->ts_fidx + l);
```
Comment (zeertzjq) [inline src/nvim/spell.c:4268]: This is tricky. An overflow can indeed happen here when there are too many composing characters.

File: src/nvim/spell.c [hunk 5]
```diff
@@ -3821,13 +3815,13 @@ static void suggest_trie_walk(suginfo_T *su, langp_T *lp, char_u *fword, bool so
       if (sp->ts_prefixdepth == PFD_PREFIXTREE) {
         // Skip over the NUL bytes, we use them later.
         for (n = 0; n < len && byts[arridx + n] == 0; n++) {}
-        sp->ts_curi += n;
+        sp->ts_curi = (int16_t)(sp->ts_curi + n);
```
[the remark this answers is not in the document]
Comment (dundargoc) [reply src/nvim/spell.c]: No, this will create a conversion warning.

File: src/nvim/spell.c (hunk 4 above)
Comment (dundargoc) [reply src/nvim/spell.c:4268]: Should I add an assert or...?

Comment (zeertzjq) [reply src/nvim/spell.c:4268]: I don't think so.

**Pull request merged**